c_src/quicer_listener.c

```diff
@@ -465,21 +463,140 @@ close_listener1(ErlNifEnv *env,
                 const ERL_NIF_TERM argv[])
 {
   QuicerListenerCTX *l_ctx;
+  BOOLEAN is_destroy = FALSE;
+  ERL_NIF_TERM ret = ATOM_OK;
   if (!enif_get_resource(env, argv[0], ctx_listener_t, (void **)&l_ctx))
     {
       return ERROR_TUPLE_2(ATOM_BADARG);
     }
 
   enif_mutex_lock(l_ctx->lock);
+  if (l_ctx->is_closed)
+    {
+      enif_mutex_unlock(l_ctx->lock);
+      return ERROR_TUPLE_2(ATOM_CLOSED);
+    }
   HQUIC l = l_ctx->Listener;
+  // set before destroy_l_ctx
   l_ctx->Listener = NULL;
   l_ctx->is_closed = TRUE;
+
+  // If is_stopped, it means the listener is already stopped.
+  // there will be no callback for QUIC_LISTENER_EVENT_STOP_COMPLETE
+  // so we need to destroy the l_ctx otherwise it will leak.
+  is_destroy = l_ctx->is_stopped;
+
   enif_mutex_unlock(l_ctx->lock);
 
-  // It is safe to close it without holding the lock
-  // This also ensures no ongoing listener callbacks
-  // This is a blocking call. @TODO have async version or use dirty scheduler
   MsQuic->ListenerClose(l);
+  if (is_destroy)
+    {
+      destroy_l_ctx(l_ctx);
+    }
+  return ret;
+}
+
+ERL_NIF_TERM
+stop_listener1(ErlNifEnv *env,
+               __unused_parm__ int argc,
+               const ERL_NIF_TERM argv[])
+{
+  QuicerListenerCTX *l_ctx;
+  ERL_NIF_TERM ret = ATOM_OK;
+  BOOLEAN is_stopped = FALSE;
+  assert(argc == 1);
+  if (!enif_get_resource(env, argv[0], ctx_listener_t, (void **)&l_ctx))
+    {
+      return ERROR_TUPLE_2(ATOM_BADARG);
+    }
+
+  enif_mutex_lock(l_ctx->lock);
+  HQUIC l = l_ctx->Listener;
+  is_stopped = l_ctx->is_stopped;
+  l_ctx->is_stopped = TRUE;
+  enif_mutex_unlock(l_ctx->lock);
+  if (!l)
+    {
+      return ERROR_TUPLE_2(ATOM_CLOSED);
+    }
+  else if (!is_stopped)
+    {
+      // void return
+      MsQuic->ListenerStop(l);
+    }
+  return ret;
+}
+
+ERL_NIF_TERM
+start_listener3(ErlNifEnv *env,
+                __unused_parm__ int argc,
+                const ERL_NIF_TERM argv[])
+{
+  ERL_NIF_TERM listener_handle = argv[0];
+  ERL_NIF_TERM elisten_on = argv[1];
+  ERL_NIF_TERM options = argv[2];
```

Comment:
Q: should we `assert(argc == 3)` before accessing those values?

Reply:
could be done. I will add it in next PR. 
btw, the assertion in this project applies to TEST build only.